README.md
# About

This repository contains a Proof of Concept for a versioning system hosted on Github.

## Building Blocks of the Solution

The primary component of this system is:

- `.github/workflows/versioning.yml`: A Github Action workflow that is triggered when a pull request is closed.

Other notable components include:

- `.github/pull_request_template.md`: A template that is automatically applied when creating a new pull request.
- `helpers/commit-message-by-ai.sh`: A script that suggests a commit message based on the git diff. It's designed for macOS; Linux compatibility hasn't been tested.
- `helpers/pr-message-by-ai.sh`: A script that suggests a pull request message based on the git diff. Like the previous script, it's tailored for macOS, with Linux compatibility untested.

## Workflow Overview

1. Start with a main branch (e.g., `master`). When creating a new branch, use a prefix like `bug/`, `task/`, or `feat/`.
2. After coding, consider using the commit message suggestion script: `./helpers/commit-message-by-ai.sh "YOUR_OPENAI_API_KEY"`. If the suggested message isn't satisfactory, run the script again or craft your own message.
3. Once you've committed your changes, open a pull request on github.com, e.g., merging `bug/TASK-123_Exchanging_x_with_y` into `master`. The pull request template will be applied automatically.
4. If you wish to modify the pull request message, use the script: `./pr-message-by-ai.sh "master" "bug/TASK-123_Exchanging_x_with_y" "YOUR_OPENAI_API_KEY"`. This will suggest you the message which you may use or not.
5. Upon closing the pull request, the `.github/workflows/versioning.yml` is triggered, leading to:
    - The latest version is determined based on the branch prefix:
      - **major**: major/, release/
      - **minor**: minor/, feature/, feat/, task/, refactor/, poc/, innovation/
      - **patch**: patch/, bug/, doc/, tools/, hotfix/, **any other**
    - An update to `version.txt` with the latest version, adhering to Semantic Versioning rules.
    - An update to `CHANGELOG.md` with the latest changes. Version number and Initial pull request title goes as a content.
    - An update to `database` directory - moving `database\update.sql` and `database\rollback.sql` files to appropriate version dir and appending one extra query to each.
    - Creation of a git tag corresponding to the new version.
    - Pushing `version.txt`, `CHANGELOG.md`, and the new tag to the remote git repository.

## Important Notes

- Ensure "Read and write permissions" are enabled for Github Actions in your repository (Settings > Actions > General > Workflow permissions).
- Generate a Personal Access Token with `repo` scope (Settings > Developer settings > Personal access tokens).
- Add the token as a secret in your repository settings with the key "GH_PAT" (Settings > Secrets and variables > Actions > Secrets > Repository secrets).
- Ensure your OpenAI account has sufficient funds for API calls. Note that a Plus subscription to ChatGPT isn't equivalent. Generate a valid OpenAI API key in your OpenAI account settings (https://platform.openai.com/account/api-keys).

## Disclaimer

The helper scripts ([commit-message-by-ai.sh](helpers%2Fcommit-message-by-ai.sh) and [pr-message-by-ai.sh](helpers%2Fpr-message-by-ai.sh)) utilize OpenAI for suggestions. Be cautious: sharing code with the OpenAI API might pose a security risk, especially if proprietary information is involved. Use these scripts at your discretion and risk.


CHANGELOG.md
# Changelog

## 8.336.1
* Update README.md


## 8.336.0
* IR-123 DB versioning solution. Solution testing.


## 8.335.4
* IR-123 DB versioning solution. Solution tweaking.


## 8.335.3
* Bug/ir 123 bug fix this


## 8.335.2
* Bug/ir 123 bug fix this


## 8.335.1
* IR-123 Did the change. PRt1


## 8.335.0
* Task/poc 9 my random task. PRt1


## 8.334.0
* POC-9 Updating README.second.md PRt1


## 8.333.3
* Bug/poc 8 some bug name. t1


## 8.333.2
* POC-8 pull_request_template.md moved to another default dir. PRt1


## 8.333.1
* POC-8 README.second.md change like a bugfix. PRt1


## 8.333.0
* Task/poc 9 my random task. PRt1


## 8.332.0
* POC-9 Some fixes PRt1


## 8.331.0
* POC-9 changing the way version bump is achieved PR t1


## 8.330.7
* POC-9 adding feat/ and some debug info. PR t1


## 8.330.6
* POC-9 Adding PR template. Title. PR title1.


## 8.330.5
* POC-9 This line is added in the scope of a task. PR title1


## 8.330.4
* POC-8 This line is added in the scope of a bug. PR title1


## 8.330.3
* Patch/poc 7 PR title1


## 8.330.2
* POC-7 Revert changes to fix version inconsistency. PR title1


## 8.330.1
* One more change. (POC-7)

## 8.330.0
* Sixth change (POC-5)

## 8.329.1
* Fifth change (POC-5)

## 8.329.0
* Second change (POC-2)

## 8.328.4
* Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. (IR-1016)

## 8.328.3
* Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. (IR-1016)

## 8.328.2
* Fix: Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. (IR-854)

## 8.328.1
* Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. (IR-918)

## 8.328.0
* Sapien eget mi proin sed libero enim sed faucibus.  (IR-872)

## 8.327.1
* Fix: Id interdum velit laoreet id donec ultrices tincidunt arcu non (SE-1045)

## 8.327.0
* Mauris pellentesque pulvinar pellentesque habitant (IR-1016)

## 8.326.2
* Fix: Nibh mauris cursus mattis molestie a iaculis. (IR-918)

## 8.326.1
* Pretium vulputate sapien nec sagittis aliquam (SE-1045)

## 8.326.0
* Aliquam id diam maecenas ultricies mi eget mauris pharetra. Lobortis scelerisque fermentum dui faucibus in ornare.
* Amet consectetur adipiscing elit pellentesque habitant morbi tristique senectus et (IR-1036)

## 8.325.0
* Luctus venenatis lectus magna fringilla (IR-1017)

## 8.324.0
* Proin libero nunc consequat interdum varius sit amet mattis (IR-626)

## 8.323.1
* Fix Massa tincidunt nunc pulvinar sapien. (IR-841)



helpers/commit-message-by-ai.sh
#!/bin/sh

OPENAI_API_KEY="$1"
OPENAI_ENGINE="gpt-3.5-turbo-instruct-0914"

set -e

echo "Versions check (script stops if any of these fail):"
git --version
curl --version
jq --version

# Generate git diff
git diff && git diff --cached > './.temp/git_diff_output.txt'

# Check if diff_output.txt is empty
if [[ ! -s './.temp/git_diff_output.txt' ]]; then
    echo "No differences found."
    exit 0
fi

if [[ -z "$OPENAI_API_KEY" ]]; then
    echo "Error: OpenAI API key not set."
    exit 1
fi

# Create a JSON payload using jq
PAYLOAD=$(jq -n \
            --arg diff "$(cat './.temp/git_diff_output.txt')" \
            '{prompt: ("Describe the main intent of the following code change in a clear and concise manner, limit the response to 50 characters, it will be in use for git commit message, all of the following is a git diff of the change: " + $diff), max_tokens: 150}')

# Send to AI and get summary using the /completions endpoint
RESPONSE=$(curl -s -X POST -H "Authorization: Bearer $OPENAI_API_KEY" \
               -H "Content-Type: application/json" \
               -d "$PAYLOAD" \
               https://api.openai.com/v1/engines/$OPENAI_ENGINE/completions)

echo "API Response:"
echo "$RESPONSE"

# Remove newline characters from the API response using sed (macOS compatible version)
CLEANED_RESPONSE=$(echo "$RESPONSE" | sed -e ':a' -e 'N' -e '$!ba' -e 's/\n//g')

# Extract the summary using jq
SUMMARY=$(echo "$CLEANED_RESPONSE" | jq -r '.choices[0].text')

# Check if SUMMARY is valid
if [[ "$SUMMARY" == "null" || -z "$SUMMARY" ]]; then
    echo "Error: Unable to get a valid summary."
    exit 1
fi

echo "Summary of Git Diff:"
echo "$SUMMARY"

helpers/pr-message-by-ai.sh
#!/bin/sh


OPENAI_ENGINE="gpt-3.5-turbo-instruct-0914"

# Get the target and destination branch names from the script parameters
SCRIPT_NAME=$(basename "$0")
TARGET_BRANCH="$1"
DESTINATION_BRANCH="$2"
OPENAI_API_KEY="$3"

# Check if both branches are provided
if [ -z "$TARGET_BRANCH" ] || [ -z "$DESTINATION_BRANCH" ]; then
    echo "Available branches:"
    git branch -a
    echo "----------------------------------------"
    echo "Usage: ./${SCRIPT_NAME} <target_branch> <destination_branch>"
    exit 1
fi

set -e

echo "Versions check (script stops if any of these fail):"
git --version
curl --version
jq --version

# Print the branch names (or do whatever you want with them)
echo "Comparing target branch: $TARGET_BRANCH with destination branch: $DESTINATION_BRANCH"

# Generate git diff
git diff "${TARGET_BRANCH}" "${DESTINATION_BRANCH}" > './.temp/git_diff_pr_output.txt'

# Check if diff_output.txt is empty
if [[ ! -s './.temp/git_diff_pr_output.txt' ]]; then
    echo "No differences found."
    exit 0
fi

if [[ -z "$OPENAI_API_KEY" ]]; then
    echo "Error: OpenAI API key not set."
    exit 1
fi

# Create a JSON payload using jq
PAYLOAD=$(jq -n \
            --arg diff "$(cat './.temp/git_diff_pr_output.txt')" \
            '{prompt: ("Describe the main intent of the following code change in a clear and concise manner, limit the response to 200 characters, it will be in use for pull request message, all of the following is a git diff of the change: " + $diff), max_tokens: 150}')

# Send to AI and get summary using the /completions endpoint
RESPONSE=$(curl -s -X POST -H "Authorization: Bearer $OPENAI_API_KEY" \
               -H "Content-Type: application/json" \
               -d "$PAYLOAD" \
               https://api.openai.com/v1/engines/$OPENAI_ENGINE/completions)

echo "API Response:"
echo "$RESPONSE"

# Remove newline characters from the API response using sed (macOS compatible version)
CLEANED_RESPONSE=$(echo "$RESPONSE" | sed -e ':a' -e 'N' -e '$!ba' -e 's/\n//g')

# Extract the summary using jq
SUMMARY=$(echo "$CLEANED_RESPONSE" | jq -r '.choices[0].text')

# Check if SUMMARY is valid
if [[ "$SUMMARY" == "null" || -z "$SUMMARY" ]]; then
    echo "Error: Unable to get a valid summary."
    exit 1
fi

echo "Summary of Git Diff:"
echo "$SUMMARY"